ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/PullQueryPublisher.java

```diff
@@ -42,32 +45,50 @@
 
 class PullQueryPublisher implements Flow.Publisher<Collection<StreamedRow>> {
 
+  private final KsqlEngine ksqlEngine;
   private final ServiceContext serviceContext;
   private final ConfiguredStatement<Query> query;
   private final PullQueryExecutor pullQueryExecutor;
   private final Optional<PullQueryExecutorMetrics> pullQueryMetrics;
   private final long startTimeNanos;
+  private final RoutingFilterFactory routingFilterFactory;
+  private final RateLimiter rateLimiter;
 
   @VisibleForTesting
   PullQueryPublisher(
+      final KsqlEngine ksqlEngine,
       final ServiceContext serviceContext,
       final ConfiguredStatement<Query> query,
       final PullQueryExecutor pullQueryExecutor,
       final Optional<PullQueryExecutorMetrics> pullQueryMetrics,
-      final long startTimeNanos
+      final long startTimeNanos,
+      final RoutingFilterFactory routingFilterFactory,
+      final RateLimiter rateLimiter
   ) {
+    this.ksqlEngine = requireNonNull(ksqlEngine, "ksqlEngine");
     this.serviceContext = requireNonNull(serviceContext, "serviceContext");
     this.query = requireNonNull(query, "query");
     this.pullQueryExecutor = requireNonNull(pullQueryExecutor, "pullQueryExecutor");
     this.pullQueryMetrics = pullQueryMetrics;
     this.startTimeNanos = startTimeNanos;
+    this.routingFilterFactory = requireNonNull(routingFilterFactory, "routingFilterFactory");
+    this.rateLimiter = requireNonNull(rateLimiter, "rateLimiter");
   }
 
   @Override
   public synchronized void subscribe(final Subscriber<Collection<StreamedRow>> subscriber) {
     final PullQuerySubscription subscription = new PullQuerySubscription(
         subscriber,
         () -> {
+          ksqlEngine.executePullQuery(
```

Comment:
uh - shouldn't we be using the result of this? otherwise we're just executing the pull query twice, right? once here and once on line 92?

also now that this refactor is done, can we get rid of `PullQueryExecutor`?

Reply:
Yes, I still need to remove all the `PullQueryExecutor` code. I am leaving it for last because it helps with debugging to verify what the correct behavior should be.